nexus/reconfigurator/execution/src/datasets.rs

```diff
@@ -400,16 +376,17 @@ mod tests {
                 .expect("failed to upsert zpool");
         }
 
-        // Call `ensure_dataset_records_exist` again, adding new datasets.
+        // Call `ensure_crucible_dataset_records_exist` again, adding new
+        // datasets.
         //
         // It should only insert these new zones.
         let new_zones = [
```

Comment:
This seems wrong, though I haven't looked deeply into this test: it's inserting two of the same zone that are different types than before this diff.

Reply:
I think this is okay, but please double check me. Before, the execution code would insert dataset records for any `DatasetKind` since `dataset` held all dataset kinds; now, it only inserts records for Crucible datasets. The test is trying to ensure that new datasets added after the last execution are added, which is now restricted to "new _Crucible_ datasets added after the last execution are added", so the new zones we use to test have to be Crucible zones.

Reply:
> Before, the execution code would insert dataset records for any DatasetKind since dataset held all dataset kinds; now, it only inserts records for Crucible datasets.

I think this is why I was suspicious: based on that understanding this vec should have had four members, mixing together the crucible and non-crucible datasets.

There should be two versions of this test: one for "durable" datasets and one for crucible datasets.

Reply:
I might be missing something - does it help if `all_datasets` already included Crucible datasets? (Guaranteed by https://github.com/oxidecomputer/omicron/pull/7386/files/26ccf5f2e0f22721cd4d1d26bac4d9c4dc659e6b#diff-d4165f273b361528feca0c934574df6365cd86f5688ab81665f35975620183a1R312-R316, and then we pass in `all_datasets.iter().chain(new_zones)`; that used to provide a mix of crucible and non-crucible, and now just passes in "more crucible".)

Reply:
We chatted about what made this test confusing; I took a stab at addressing it in 1201f9a8c